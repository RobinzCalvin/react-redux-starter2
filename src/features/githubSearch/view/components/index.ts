export { default as UserAvatarsWall } from './UserAvatarsWall/UserAvatarsWall';
export { default as SearchSettingsDialog } from './SearchSettingsDialog/SearchSettingsDialog';
