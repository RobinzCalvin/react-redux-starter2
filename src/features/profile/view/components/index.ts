export { default as ProfileAvatar } from './ProfileAvatar/ProfileAvatar';
