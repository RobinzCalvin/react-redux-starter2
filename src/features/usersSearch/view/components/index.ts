export { default as UserAvatarsWall } from './UserAvatarsWall/UserAvatarsWall';
export { default as UserAttribute } from './UserAttribute/UserAttribute';
