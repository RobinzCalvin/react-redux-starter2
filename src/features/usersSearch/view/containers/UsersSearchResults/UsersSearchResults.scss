.users-search-results {
  &__pagination {
    margin-top: 1rem;
  }
}
