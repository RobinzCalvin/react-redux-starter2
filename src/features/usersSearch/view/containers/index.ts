export { default as UsersSearchForm } from './UsersSearchForm/UsersSearchForm';
export { default as UsersSearchResults } from './UsersSearchResults/UsersSearchResults';
export { default as UserDetails } from './UserDetails/UserDetails';
