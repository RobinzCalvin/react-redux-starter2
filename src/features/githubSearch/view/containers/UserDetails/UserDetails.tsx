import * as React from 'react';
import block from 'bem-cn';
import { bind } from 'decko';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { IAppReduxState } from 'shared/types/app';
import { IUser } from 'shared/types/models';
import { Dialog } from 'shared/view/components';

import { actions, selectors } from './../../../redux';
import './UserDetails.scss';

interface IOwnProps {
  userLogin: string | null; // think of login or username everywhere
  onClose(): void;
}

interface IStateProps {
  userDetails: IUser | null;
}

interface IActionProps {
  loadUserDetails: typeof actions.loadUserDetails;
}

type IProps = IStateProps & IActionProps & IOwnProps;

function mapState(state: IAppReduxState): IStateProps {
  return {
    userDetails: selectors.selectUserDetails(state),
  };
}

function mapDispatch(dispatch: Dispatch): IActionProps {
  return bindActionCreators({
    loadUserDetails: actions.loadUserDetails,
  }, dispatch);
}

const b = block('user-details');

class UserDetails extends React.PureComponent<IProps> {
  public render() {
    const { userLogin } = this.props;
    return (
      <Dialog
        open={userLogin !== null}
        title="User details"
        onEnter={this.handleDialogEnter}
        onClose={this.handleDialogClose}
      >
        <div className={b()}>content</div>
      </Dialog>
    );
  }

  @bind
  private handleDialogEnter() {
    const { userLogin, loadUserDetails } = this.props;
    if (userLogin) {
      loadUserDetails(userLogin);
    }
  }

  @bind
  private handleDialogClose() {
    const { onClose } = this.props;
    onClose();
  }
}

export default connect(mapState, mapDispatch)(UserDetails);
