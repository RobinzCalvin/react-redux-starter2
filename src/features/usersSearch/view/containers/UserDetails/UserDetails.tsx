import * as React from 'react';
import block from 'bem-cn';
import { bind } from 'decko';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { IAppReduxState } from 'shared/types/app';
import { IDetailedUser } from 'shared/types/models';
import { Dialog } from 'shared/view/components';
import { Typography, Preloader } from 'shared/view/elements';

import { injectSizeToAvatarURL } from '../../../helpers';
import { actions, selectors } from './../../../redux';
import './UserDetails.scss';

interface IOwnProps {
  username: string;
  onClose(): void;
}

interface IStateProps {
  userDetails: IDetailedUser | null;
  isLoadUserDetailsRequesting: boolean;
}

interface IActionProps {
  loadUserDetails: typeof actions.loadUserDetails;
}

type IProps = IStateProps & IActionProps & IOwnProps;

function mapState(state: IAppReduxState): IStateProps {
  return {
    userDetails: selectors.selectUserDetails(state),
    isLoadUserDetailsRequesting: selectors.selectCommunication(state, 'loadUserDetails').isRequesting,
  };
}

function mapDispatch(dispatch: Dispatch): IActionProps {
  return bindActionCreators({
    loadUserDetails: actions.loadUserDetails,
  }, dispatch);
}

const b = block('user-details');

class UserDetails extends React.PureComponent<IProps> {
  private avatarSize = 230;

  public render() {
    const { isLoadUserDetailsRequesting } = this.props;
    return (
      <Dialog
        open={true}
        title="User details"
        onEnter={this.handleDialogEnter}
        onClose={this.handleDialogClose}
      >
        <div className={b()}>
          <Preloader size={80} isShown={isLoadUserDetailsRequesting} backgroundColor="#fff" />
          {this.renderContent()}
        </div>
      </Dialog>
    );
  }

  private renderContent() {
    const { userDetails } = this.props;
    if (userDetails) {
      return (
        <>
          {this.renderMain(userDetails)}
          {this.renderAttributes(userDetails)}
        </>
      );
    }
    return null;
  }

  private renderMain(userDetails: IDetailedUser) {
    const { htmlURL, avatarURL, realName, username, location } = userDetails;
    return (
      <a href={htmlURL} className={b('main')} target="_blank">
        <img
          className={b('avatar')}
          src={injectSizeToAvatarURL(avatarURL, this.avatarSize)}
        />
        <Typography variant="h5">{realName}</Typography>
        <Typography variant="subtitle1">{username}</Typography>
        {location && <Typography variant="subtitle2" color="textSecondary">{location}</Typography>}
      </a>
    );
  }

  private renderAttributes(userDetails: IDetailedUser) {
    const { followersNumber, followingNumber, reposNumber, htmlURL } = userDetails;
    return (
      <div className={b('attributes')}>
        {this.renderAttribute(`${htmlURL}/followers`, 'Followers', followersNumber)}
        {this.renderAttribute(`${htmlURL}/following`, 'Following', followingNumber)}
        {this.renderAttribute(`${htmlURL}/repositories`, 'Repositories', reposNumber)}
      </div>
    );
  }

  private renderAttribute(URL: string, title: string, value: number) {
    return (
      <a href={URL} target="_blank" className={b('attribute')}>
        {title}
        <span className={b('value')}>{value}</span>
      </a>
    );
  }

  @bind
  private handleDialogEnter() {
    const { username, loadUserDetails } = this.props;
    if (username) {
      loadUserDetails(username);
    }
  }

  @bind
  private handleDialogClose() {
    const { onClose } = this.props;
    onClose();
  }
}

export default connect(mapState, mapDispatch)(UserDetails);
