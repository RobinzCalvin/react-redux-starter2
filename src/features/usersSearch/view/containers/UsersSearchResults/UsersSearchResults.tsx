import React from 'react';
import { connect } from 'react-redux';
import block from 'bem-cn';
import { bind } from 'decko';

import { IAppReduxState } from 'shared/types/app';
import { IGithubUser } from 'shared/types/models';
import { IPaginationState } from 'shared/types/common';
import { PaginationControls } from 'shared/view/components';

import { IUsersSearchFormFields } from '../../../namespace';
import { UserAvatarsWall } from '../../components';
import { actions, selectors } from './../../../redux';
import UserDetails from '../UserDetails/UserDetails';
import './UsersSearchResults.scss';

interface IState {
  displayedUser: string | null;
}

interface IOwnProps {
  searchOptions: IUsersSearchFormFields;
}

interface IStateProps {
  users: IGithubUser[];
  paginationState: IPaginationState;
}

type IActionProps = typeof mapDispatch;

type IProps = IOwnProps & IStateProps & IActionProps;

function mapState(state: IAppReduxState): IStateProps {
  return {
    users: selectors.selectFoundUsers(state),
    paginationState: selectors.selectUsersSearchPaginationState(state),
  };
}

const mapDispatch = {
  searchUser: actions.searchUser,
};

const b = block('users-search-results');

class UsersSearchResults extends React.PureComponent<IProps> {
  public state: IState = {
    displayedUser: null,
  };

  public render() {
    const { users, paginationState: { page, totalPages } } = this.props;
    const { displayedUser } = this.state;
    return (
      <div className={b()}>
        <UserAvatarsWall users={users} onAvatarClick={this.handleUserAvatarClick} />
        <div className={b('pagination')}>
          <PaginationControls
            totalPages={totalPages}
            currentPage={page}
            onPageRequest={this.handlePageRequest}
          />
        </div>
        {displayedUser && <UserDetails username={displayedUser} onClose={this.handleUserDetailsClose}/>}
      </div>
    );
  }

  @bind
  private handlePageRequest(pageNumber: number) {
    const { searchUser, searchOptions } = this.props;
    searchUser({ searchOptions, page: pageNumber });
  }

  @bind
  private handleUserAvatarClick({ username }: IGithubUser) {
    this.setState({ displayedUser: username });
  }

  @bind
  private handleUserDetailsClose() {
    this.setState({ displayedUser: null });
  }
}

export default connect(mapState, mapDispatch)(UsersSearchResults);
