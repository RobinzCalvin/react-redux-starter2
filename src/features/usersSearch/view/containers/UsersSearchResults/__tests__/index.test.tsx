import React from 'react';
import { shallow } from 'enzyme';

import { githubUser } from 'shared/mocks';
import { PaginationControls } from 'shared/view/components';

import { UsersAvatarsWall } from '../../../components';
import UserDetails from '../../UserDetails/UserDetails';
import { UsersSearchResults, IUsersSearchResultsProps } from '../UsersSearchResults';

const props: IUsersSearchResultsProps = {
  searchOptions: {
    perPage: 30,
    searchBy: 'email',
    searchString: 'search',
    searchType: 'user',
  },
  users: Array(10).fill(githubUser),
  paginationState: {
    page: 1,
    totalPages: 2,
  },
  searchUsers: jest.fn(),
};

describe('(features/usersSearch) UsersSearchResults container', () => {
  const component = shallow(<UsersSearchResults {...props} />);

  it('should show UserDetails on user avatar click', () => {
    expect(component.find(UserDetails).length).toBe(0);

    component.find(UsersAvatarsWall).prop('onAvatarClick')(githubUser);
    expect(component.find(UserDetails).length).toBe(1);
  });

  it('should call searchUser with search options and page number on PaginationControls page request', () => {
    const page = 1;
    const { searchOptions, searchUsers } = props;
    component.find(PaginationControls).prop('onPageRequest')(page);
    expect(searchUsers).toHaveBeenCalledWith({ searchOptions, page });
  });
});
