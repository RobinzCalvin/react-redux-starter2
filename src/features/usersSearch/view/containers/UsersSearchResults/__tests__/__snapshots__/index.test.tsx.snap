// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersSearchResults component should match snapshot 1`] = `
<div
  className="users-search-results"
>
  <UsersAvatarsWall
    onAvatarClick={[Function]}
    users={
      Array [
        Object {
          "avatarURL": "http://the-url.com",
          "htmlURL": "http://the-url.com",
          "id": 1,
          "username": "username",
        },
        Object {
          "avatarURL": "http://the-url.com",
          "htmlURL": "http://the-url.com",
          "id": 1,
          "username": "username",
        },
        Object {
          "avatarURL": "http://the-url.com",
          "htmlURL": "http://the-url.com",
          "id": 1,
          "username": "username",
        },
        Object {
          "avatarURL": "http://the-url.com",
          "htmlURL": "http://the-url.com",
          "id": 1,
          "username": "username",
        },
        Object {
          "avatarURL": "http://the-url.com",
          "htmlURL": "http://the-url.com",
          "id": 1,
          "username": "username",
        },
        Object {
          "avatarURL": "http://the-url.com",
          "htmlURL": "http://the-url.com",
          "id": 1,
          "username": "username",
        },
        Object {
          "avatarURL": "http://the-url.com",
          "htmlURL": "http://the-url.com",
          "id": 1,
          "username": "username",
        },
        Object {
          "avatarURL": "http://the-url.com",
          "htmlURL": "http://the-url.com",
          "id": 1,
          "username": "username",
        },
        Object {
          "avatarURL": "http://the-url.com",
          "htmlURL": "http://the-url.com",
          "id": 1,
          "username": "username",
        },
        Object {
          "avatarURL": "http://the-url.com",
          "htmlURL": "http://the-url.com",
          "id": 1,
          "username": "username",
        },
      ]
    }
  />
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <PaginationControls
      currentPage={1}
      onPageRequest={[Function]}
      totalPages={2}
    />
  </div>
</div>
`;
